walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,206 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    private static ArrayList<Event> optionalEvents = new ArrayList<>();
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> times = new ArrayList<>();
+
+        if ((int) request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return times;
+        } else if (events.size() == 0) {
+            times.add(TimeRange.WHOLE_DAY);
+
+            return times;
+        } else if (events.size() == 1) {
+            boolean included = false;
+
+            for (Event event : events) {
+                if (attends(event, request)) {
+                    included = true;
+                }
+            }
+
+            if (included) {
+                Event[] e = events.toArray(new Event[0]);
+                TimeRange time = e[0].getWhen();
+
+                if (time.start() - TimeRange.START_OF_DAY >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, time.start(), false));
+                }
+
+                if (TimeRange.END_OF_DAY - time.end() >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(time.end(), TimeRange.END_OF_DAY, true));
+                }
+            } else {
+                times.add(TimeRange.WHOLE_DAY);
+            }
+
+            return times;
+        } else {
+            Event[] e = events.toArray(new Event[0]);
+            Event[] sortedEvents = insertionSort(e, TimeRange.ORDER_BY_START, request);;
+
+            if (sortedEvents.length == 0) {
+                if (optionalEvents.size() == 0) {
+                    times.add(TimeRange.WHOLE_DAY);
+                }
+        
+                return times;
+            } else if (sortedEvents.length == 1) {
+                TimeRange time = sortedEvents[0].getWhen();
+
+                if (time.start() - TimeRange.START_OF_DAY >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, time.start(), false));
+                }
+
+                if (TimeRange.END_OF_DAY - time.end() >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(time.end(), TimeRange.END_OF_DAY, true));
+                }
+
+                return times;
+            } else {
+                int startTime = TimeRange.START_OF_DAY;
+                int endTime = TimeRange.END_OF_DAY;
+
+                TimeRange first = sortedEvents[0].getWhen();
+                TimeRange second = sortedEvents[1].getWhen();
+
+                int tempStart = startTime;
+                int tempEnd = endTime;
+
+                boolean optional = false;
+                int previous = startTime;
+
+                if (first.overlaps(second)) {
+                    tempEnd = Math.min(first.start(), second.start());
```

Comment:
Since the events are sorted by their start times tempEnd will always be first.start() irrespective of whether first and second overlap.  (This is based on my guess of what you intend tempEnd to be because the name 'tempEnd' is not very descriptive.)